config/auth_u2m.go

```diff
@@ -0,0 +1,198 @@
+// error.
+type U2MCredentials struct {
+	// Auth is the persistent auth object to use. If not specified, a new one will
+	// be created, using the default cache and locker.
+	Auth *oauth.PersistentAuth
+
+	// GetOAuthArg is a function that returns the OAuth argument to use for
+	// loading the OAuth session token. If not specified, the OAuth argument is
+	// determined by the account host and account ID or workspace host in the
+	// Config.
+	GetOAuthArg func(context.Context, *Config) (oauth.OAuthArgument, error)
+
+	// ErrorHandler controls the behavior of Configure() when loading the OAuth
+	// token fails. If not specified, any error will cause Configure() to return
+	// said error.
+	ErrorHandler func(context.Context, *Config, oauth.OAuthArgument, error) error
+
+	name string
+}
+
+// Name implements CredentialsStrategy.
+func (u U2MCredentials) Name() string {
+	if u.name != "" {
+		return "oauth-u2m"
+	}
+	return u.name
+}
+
+// Configure implements CredentialsStrategy.
+func (u U2MCredentials) Configure(ctx context.Context, cfg *Config) (credentials.CredentialsProvider, error) {
+	a := u.Auth
+	if a == nil {
+		var err error
+		a, err = oauth.NewPersistentAuth(ctx)
+		if err != nil {
+			logger.Debugf(ctx, "failed to create persistent auth: %v, continuing", err)
+			return nil, nil
+		}
+	}
+
+	var arg oauth.OAuthArgument
+	var err error
+	if u.GetOAuthArg != nil {
+		arg, err = u.GetOAuthArg(ctx, cfg)
+	} else {
+		arg, err = defaultGetOAuthArg(ctx, cfg)
+	}
+	if err != nil {
+		return nil, fmt.Errorf("oidc: %w", err)
+	}
+
+	r, err := http.NewRequestWithContext(ctx, http.MethodGet, "", nil)
+	if err != nil {
+		return nil, fmt.Errorf("http request: %w", err)
+	}
+
+	f := func(r *http.Request) error {
+		token, err := a.Load(r.Context(), arg)
+		if err != nil {
+			return fmt.Errorf("oidc: %w", err)
+		}
+		r.Header.Set("Authorization", fmt.Sprintf("Bearer %s", token.AccessToken))
+		return nil
+	}
+
+	// Try to load the credential from the token cache. If absent, fall back to
+	// the next credentials strategy. If a token is present but cannot be loaded
+	// (e.g. expired), return an error. Otherwise, fall back to the next
+	// credentials strategy.
+	if err := f(r); err != nil {
+		if u.ErrorHandler != nil {
+			return nil, u.ErrorHandler(ctx, cfg, arg, err)
+		}
+		return nil, err
+	}
+
+	return credentials.NewCredentialsProvider(f), nil
+}
+
+func defaultGetOAuthArg(_ context.Context, cfg *Config) (oauth.OAuthArgument, error) {
+	if cfg.IsAccountClient() {
+		return oauth.NewBasicAccountOAuthArgument(cfg.Host, cfg.AccountID)
+	}
+	return oauth.NewBasicWorkspaceOAuthArgument(cfg.Host)
+}
+
+var _ CredentialsStrategy = U2MCredentials{}
+
+// CliInvalidRefreshTokenError is a special error type that is returned when a
+// new access token could not be retrieved because the refresh token is invalid.
+// It is returned only by the `databricks-cli` credentials strategy.
```

Comment:
Because SDK users may reuse a session where they authenticated with the CLI, there is a slight cyclic dependency here. It isn't a code dependency, rather a process one: users may need to use the CLI to reauthenticate if a refresh token is invalid. Thus, the SDK will now include a small bit of logic to generate the command to reauthenticate, which can be logged to instruct the user to reauthenticate.